In [1]:
import pandas as pd


import matplotlib.pyplot as plt
plt.style.use('ggplot')
#%matplotlib inline  

from sklearn import linear_model
from sklearn import model_selection
from sklearn import metrics
from sklearn.ensemble import RandomForestRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.feature_selection import RFECV


In [3]:
# load training data prepared previously

df_train = pd.read_csv('train.csv')
df_train.head()


id  cycle  setting1  setting2  setting3      s1      s2       s3       s4  \
0   1      1   -0.0007   -0.0004     100.0  518.67  641.82  1589.70  1400.60   
1   1      2    0.0019   -0.0003     100.0  518.67  642.15  1591.82  1403.14   
2   1      3   -0.0043    0.0003     100.0  518.67  642.35  1587.99  1404.20   
3   1      4    0.0007    0.0000     100.0  518.67  642.35  1582.79  1401.87   
4   1      5   -0.0019   -0.0002     100.0  518.67  642.37  1582.85  1406.22   

      s5    ...          sd15  sd16      sd17  sd18  sd19      sd20      sd21  \
0  14.62    ...      0.000000   0.0  0.000000   0.0   0.0  0.000000  0.000000   
1  14.62    ...      0.008697   0.0  0.000000   0.0   0.0  0.042426  0.003253   
2  14.62    ...      0.007640   0.0  1.154701   0.0   0.0  0.055076  0.044573   
3  14.62    ...      0.028117   0.0  1.000000   0.0   0.0  0.076322  0.037977   
4  14.62    ...      0.025953   0.0  1.095445   0.0   0.0  0.073621  0.033498   

   ttf  label_bnc  label_mcc  
0  191          0          0  
1  190          0          0  
2  189          0          0  
3  188          0          0  
4  187          0          0  

[5 rows x 71 columns]

In [4]:
# load test data prepared previously

df_test = pd.read_csv('test.csv')
df_test.head()

id  cycle  setting1  setting2  setting3      s1      s2       s3       s4  \
0   1     31   -0.0006    0.0004     100.0  518.67  642.58  1581.22  1398.91   
1   2     49    0.0018   -0.0001     100.0  518.67  642.55  1586.59  1410.83   
2   3    126   -0.0016    0.0004     100.0  518.67  642.88  1589.75  1418.89   
3   4    106    0.0012    0.0004     100.0  518.67  642.78  1594.53  1406.88   
4   5     98   -0.0013   -0.0004     100.0  518.67  642.27  1589.94  1419.36   

      s5    ...          sd15  sd16      sd17  sd18  sd19      sd20      sd21  \
0  14.62    ...      0.021174   0.0  0.707107   0.0   0.0  0.124016  0.025037   
1  14.62    ...      0.011429   0.0  1.095445   0.0   0.0  0.069785  0.027820   
2  14.62    ...      0.030924   0.0  0.836660   0.0   0.0  0.143631  0.018966   
3  14.62    ...      0.021249   0.0  1.341641   0.0   0.0  0.126372  0.021106   
4  14.62    ...      0.009627   0.0  0.547723   0.0   0.0  0.091924  0.101100   

   ttf  label_bnc  label_mcc  
0  112          0          0  
1   98          0          0  
2   69          0          0  
3   82          0          0  
4   91          0          0  

[5 rows x 71 columns]

In [20]:
# original features
features_orig = ['setting1','setting2','setting3','s1','s2','s3','s4','s5','s6','s7','s8','s9','s10','s11','s12','s13','s14','s15','s16','s17','s18','s19','s20','s21']

# original + extracted fetures
features_extr = ['setting1', 'setting2', 'setting3', 's1', 's2', 's3', 's4', 's5', 's6', 's7', 's8', 's9', 's10', 's11', 's12', 's13', 's14', 's15',
                 's16', 's17', 's18', 's19', 's20', 's21', 'av1', 'av2', 'av3', 'av4', 'av5', 'av6', 'av7', 'av8', 'av9', 'av10', 'av11', 'av12', 'av13', 'av14', 'av15', 'av16',
                 'av17', 'av18', 'av19', 'av20', 'av21', 'sd1', 'sd2', 'sd3', 'sd4', 'sd5', 'sd6', 'sd7', 'sd8', 'sd9', 'sd10', 'sd11', 'sd12', 'sd13', 'sd14', 'sd15', 'sd16', 'sd17', 'sd18', 'sd19', 'sd20', 'sd21']

# features with low or no correlation with regression label
features_lowcr = ['setting3', 's1', 's10', 's18','s19','s16','s5', 'setting1', 'setting2']

# features that have correlation with regression label
features_corrl = ['s2', 's3', 's4', 's6', 's7', 's8', 's9', 's11', 's12', 's13', 's14', 's15', 's17', 's20','s21']

features = features_orig


X_train = df_train[features]
y_train = df_train['ttf']

X_test = df_test[features]
y_test = df_test['ttf']

In [22]:
def get_regression_metrics(model, actual, predicted):
    
    """Calculate main parameters for regression.
    
    Args:
        model (str): The model name identifier
        actual (series): Contains the test label values
        predicted (series): Contains the predicted values
        
    Returns:
        dataframe: The combined metrics in single dataframe
    
    
    """
    regr_metrics = {
                        'Root Mean Squared Error' : metrics.mean_squared_error(actual, predicted)**0.5,
                        'Mean Absolute Error' : metrics.mean_absolute_error(actual, predicted),
                        'R^2' : metrics.r2_score(actual, predicted),
                        'Explained Variance' : metrics.explained_variance_score(actual, predicted)
                   }

    #return reg_metrics
    df_regr_metrics = pd.DataFrame.from_dict(regr_metrics, orient='index')
    df_regr_metrics.columns = [model]
    return df_regr_metrics


In [8]:
# linear regression

linreg = linear_model.LinearRegression()
linreg.fit(X_train, y_train)

y_test_predict = linreg.predict(X_test)
y_train_predict = linreg.predict(X_train)
linreg_metrics = get_regression_metrics('Linear Regression', y_test, y_test_predict)


In [9]:
# LASSO

lasso = linear_model.Lasso(alpha=0.001)
lasso.fit(X_train, y_train)

y_test_predict = lasso.predict(X_test)
y_train_predict = lasso.predict(X_train)
lasso_metrics = get_regression_metrics('LASSO', y_test, y_test_predict)


In [10]:
#try ridge

rdg = linear_model.Ridge(alpha=0.01)
rdg.fit(X_train, y_train)

y_test_predict = rdg.predict(X_test)
y_train_predict = rdg.predict(X_train)
rdg_metrics = get_regression_metrics('Ridge Regression', y_test, y_test_predict)



In [11]:
# Polynomial Regression

from sklearn.preprocessing import PolynomialFeatures

poly = PolynomialFeatures(degree=2)

X_train_poly = poly.fit_transform(X_train)
X_test_poly = poly.fit_transform(X_test)


polyreg = linear_model.LinearRegression()
polyreg.fit(X_train_poly, y_train)

y_test_predict = polyreg.predict(X_test_poly)
y_train_predict = polyreg.predict(X_train_poly)
polyreg_metrics = get_regression_metrics('Polynomial Regression', y_test, y_test_predict)


In [12]:
# Decision Tree regressor

#dtrg = DecisionTreeRegressor(max_depth=8, max_features=5, random_state=123) # selected features
dtrg = DecisionTreeRegressor(max_depth=7, random_state=123)
dtrg.fit(X_train, y_train)

y_test_predict = dtrg.predict(X_test)
y_train_predict = dtrg.predict(X_train)
dtrg_metrics = get_regression_metrics('Decision Tree Regression', y_test, y_test_predict)


In [15]:


# Random Forest

rf = RandomForestRegressor(n_estimators=100, max_features=3, max_depth=4, n_jobs=-1, random_state=1) # original features

rf.fit(X_train, y_train)

y_test_predict = rf.predict(X_test)
y_train_predict = rf.predict(X_train)
rf_metrics = get_regression_metrics('Random Forest Regression', y_test, y_test_predict)



In [16]:
reg_metrics_bfe = pd.concat([linreg_metrics, lasso_metrics, rdg_metrics, dtrg_metrics, polyreg_metrics, rf_metrics], axis=1)
reg_metrics_bfe

Linear Regression      LASSO  Ridge Regression  \
Root Mean Squared Error          32.041095  31.966099         31.965740   
Mean Absolute Error              25.591780  25.551808         25.544620   
R^2                               0.405495   0.408275          0.408289   
Explained Variance                0.665297   0.668206          0.667607   

                         Decision Tree Regression  Polynomial Regression  \
Root Mean Squared Error                 32.095349              32.262238   
Mean Absolute Error                     24.319068              24.962344   
R^2                                      0.403480               0.397261   
Explained Variance                       0.632767               0.634787   

                         Random Forest Regression  
Root Mean Squared Error                 28.634253  
Mean Absolute Error                     23.167130  
R^2                                      0.525198  
Explained Variance                       0.767320

In [17]:
# try recursive feature elimination

kfold = model_selection.KFold(n_splits=5, shuffle=True ,random_state=10)
dtrg = DecisionTreeRegressor(max_depth=7)
rfecv = RFECV(estimator=dtrg, step=1, cv=kfold, scoring='neg_mean_squared_error', n_jobs=-1)
rfecv.fit(X_train, y_train)

# Decition Tree with selected features

#trnsform train and test data
X_train_trn = rfecv.transform(X_train)
X_test_trn = rfecv.transform(X_test)


dtrg = DecisionTreeRegressor(max_depth=7)

dtrg.fit(X_train_trn, y_train)

y_test_predict = dtrg.predict(X_test_trn)
dtrg_fs_metrics = get_regression_metrics('Decision Tree: Selected Features', y_test, y_test_predict)


In [18]:
# Decition Tree with selected features

#trnsform train and test data
X_train_trn = rfecv.transform(X_train)
X_test_trn = rfecv.transform(X_test)

print(X_train.shape)

dtrg = DecisionTreeRegressor(max_depth=7)

dtrg.fit(X_train_trn, y_train)

y_test_predict = dtrg.predict(X_test_trn)

dtrg_fs_metrics = get_regression_metrics('Decision Tree: Selected Features', y_test, y_test_predict)

#combine decision tree results: All features and selected features
dtr_metrics = pd.concat([dtrg_fs_metrics,dtrg_metrics], axis=1)

dtr_metrics


(20631, 24)


Decision Tree: Selected Features  \
Root Mean Squared Error                         34.212392   
Mean Absolute Error                             25.866117   
R^2                                              0.322191   
Explained Variance                               0.593892   

                         Decision Tree Regression  
Root Mean Squared Error                 32.095349  
Mean Absolute Error                     24.319068  
R^2                                      0.403480  
Explained Variance                       0.632767

### Summary:

- As it is shown in, Random Forest performed better than other models like LASSO and Ridge regression. Random Forest clearly outperformed other models scoring RMSE of 28.63 cycles, i.e. the model predicts TTF within average error range of ±28.63 cycles. 
- The hyper-parameter tuning for the Random Forest Regressor, LASSO, and Ridge models was done manually, but Grid Search or Random Search with Cross Validation could be used for better processing. 
- Model performance using only 5 top important features is comparable to using all 24 features.